src/libraries/System.Private.CoreLib/src/System/SearchValues/Strings/StringSearchValues.cs

```diff
@@ -0,0 +1,391 @@
+        {
+            AnalyzeValues(values, ref ignoreCase, out bool allAscii, out bool asciiLettersOnly, out bool nonAsciiAffectedByCaseConversion, out int minLength);
+
+            if (values.Length == 1)
+            {
+                // We may reach this if we've removed unreachable values and ended up with only 1 remaining.
+                return CreateForSingleValue(values[0], uniqueValues, ignoreCase, allAscii, asciiLettersOnly);
+            }
+
+            // TODO: Should this be supported anywhere else?
+            // It may be too niche and too much code for WASM, but AOT with just Vector128 may be interesting.
+            if ((Ssse3.IsSupported || AdvSimd.Arm64.IsSupported) &&
+                TryGetTeddyAcceleratedValues(values, uniqueValues, ignoreCase, allAscii, asciiLettersOnly, nonAsciiAffectedByCaseConversion, minLength) is { } searchValues)
+            {
+                return searchValues;
+            }
+
+            AhoCorasick ahoCorasick = ahoCorasickBuilder.Build();
+
+            if (!ignoreCase)
+            {
+                return PickAhoCorasickImplementation<CaseSensitive>(ahoCorasick, uniqueValues);
+            }
+
+            if (nonAsciiAffectedByCaseConversion)
+            {
+                return PickAhoCorasickImplementation<CaseInsensitiveUnicode>(ahoCorasick, uniqueValues);
+            }
+
+            if (asciiLettersOnly)
+            {
+                return PickAhoCorasickImplementation<CaseInensitiveAsciiLetters>(ahoCorasick, uniqueValues);
+            }
+
+            return PickAhoCorasickImplementation<CaseInensitiveAscii>(ahoCorasick, uniqueValues);
+
+            static SearchValues<string> PickAhoCorasickImplementation<TCaseSensitivity>(AhoCorasick ahoCorasick, HashSet<string> uniqueValues)
+                where TCaseSensitivity : struct, ICaseSensitivity
+            {
+                return ahoCorasick.ShouldUseAsciiFastScan
+                    ? new StringSearchValuesAhoCorasick<TCaseSensitivity, AhoCorasick.IndexOfAnyAsciiFastScan>(ahoCorasick, uniqueValues)
+                    : new StringSearchValuesAhoCorasick<TCaseSensitivity, AhoCorasick.NoFastScan>(ahoCorasick, uniqueValues);
+            }
+        }
+
+        private static SearchValues<string>? TryGetTeddyAcceleratedValues(
+            ReadOnlySpan<string> values,
+            HashSet<string> uniqueValues,
+            bool ignoreCase,
+            bool allAscii,
+            bool asciiLettersOnly,
+            bool nonAsciiAffectedByCaseConversion,
+            int minLength)
+        {
+            if (minLength == 1)
+            {
+                // An 'N=1' implementation is possible, but callers should
+                // consider using SearchValues<char> instead in such cases.
+                // It can be added if Regex ends up running into this case.
+                return null;
+            }
+
+            if (values.Length > RabinKarp.MaxValues)
```

Comment:
For regex, how many strings would you recommend it gather to use with SearchValues?  It could gather as many prefixes as it can and always delegate to SearchValues, or it could gather some maximum number and just not use SearchValues if it's more than that.

Reply:
`["Sherlock", "Holmes", "Watson", "Irene", "Adler", "John", "Baker"]`

|             Method | WordCount |     Mean |    Error |   StdDev |
|------------------- |---------- |---------:|---------:|---------:|
| SearchValues_Count |         8 | 81.87 us | 0.412 us | 0.386 us |

As soon as you start using buckets, it's gonna really depend on the values (e.g., see SkippedTop 1000 vs 1100 numbers below).
I'd imagine bucket selection improvements would help here (but I wouldn't block this PR on that).

Sort words by frequency, skip the top X words, take Y words:
(101st word appears 79 times, 301st 29 times, 1001st 10 times)

<details>
<summary>Benchmark results</summary>

(forgive the different units on some of these - they were separate BDN runs)

`SearchValues` in this case is just
```c#
while (input not empty)
{
    int index = input.IndexOfAny(values);
    input = input.Slice(index + 1);
}
```

Regex is `new Regex(string.Join('|', words), Compiled)` + `regex.Count`

|       Method | SkippedTop | WordCount |     Mean |
|------------- |----------- |---------- |---------:|
| SearchValues |        100 |         2 | 170.1 us |
|        Regex |        100 |         2 | 596.6 us |
| SearchValues |        300 |         2 | 73.35 us |
|        Regex |        300 |         2 | 249.6 us |
| SearchValues |       1000 |         2 | 70.77 us |
|        Regex |       1000 |         2 | 315.2 us |
|              |            |           |          |
| SearchValues |        100 |         4 | 178.8 us |
|        Regex |        100 |         4 | 864.2 us |
| SearchValues |        300 |         4 | 92.25 us |
|        Regex |        300 |         4 | 561.6 us |
| SearchValues |       1000 |         4 | 74.74 us |
|        Regex |       1000 |         4 | 773.2 us |
|              |            |           |          |
| SearchValues |        100 |         8 | 219.7 us |
|        Regex |        100 |         8 | 1.319 ms |
| SearchValues |        300 |         8 | 146.0 us |
|        Regex |        300 |         8 | 759.8 us |
| SearchValues |       1000 |         8 | 94.21 us |
|        Regex |       1000 |         8 | 1.090 ms |
|              |            |           |          |
| SearchValues |        100 |        12 | 758.7 us |
|        Regex |        100 |        12 | 2.367 ms |
| SearchValues |        300 |        12 | 496.1 us |
|        Regex |        300 |        12 | 1.576 ms |
| SearchValues |       1000 |        12 | 242.6 us |
|        Regex |       1000 |        12 | 1.543 ms |
|              |            |           |          |
| SearchValues |        100 |        16 | 970.1 us |
|        Regex |        100 |        16 | 3.125 ms |
| SearchValues |        300 |        16 | 805.5 us |
|        Regex |        300 |        16 | 2.812 ms |
| SearchValues |       1000 |        16 | 253.1 us |
|        Regex |       1000 |        16 | 2.636 ms |
|              |            |           |          |
| SearchValues |        100 |        24 | 2.353 ms |
|        Regex |        100 |        24 | 4.427 ms |
| SearchValues |        300 |        24 | 1.430 ms |
|        Regex |        300 |        24 | 4.190 ms |
| SearchValues |       1000 |        24 | 5.101 ms |
|        Regex |       1000 |        24 | 5.809 ms |
| SearchValues |       1100 |        24 | 622.3 us |
|        Regex |       1100 |        24 | 5.175 ms |
|              |            |           |          |
| SearchValues |        100 |        32 | 5.519 ms |
|        Regex |        100 |        32 | 5.903 ms |
| SearchValues |        300 |        32 | 2.402 ms |
|        Regex |        300 |        32 | 4.909 ms |
| SearchValues |       1000 |        32 | 5.934 ms |
|        Regex |       1000 |        32 | 7.265 ms |
| SearchValues |       1100 |        32 | 1.134 ms |
|        Regex |       1100 |        32 | 6.218 ms |
|              |            |           |          |
| SearchValues |        100 |        40 | 5.761 ms |
|        Regex |        100 |        40 | 6.930 ms |
| SearchValues |        300 |        40 | 4.519 ms |
|        Regex |        300 |        40 | 5.767 ms |
| SearchValues |       1000 |        40 | 6.981 ms |
|        Regex |       1000 |        40 | 9.907 ms |
| SearchValues |       1100 |        40 | 1.764 ms |
|        Regex |       1100 |        40 | 6.941 ms |
|              |            |           |          |
| SearchValues |        100 |       512 | 8.274 ms |
|        Regex |        100 |       512 | 79.96 ms |
| SearchValues |        100 |      2048 | 8.599 ms |
|        Regex |        100 |      2048 | 638.5 ms |

</details>

So `<= 8` is a given, after that it depends on what the alternative is.
Something between 24-32 is probably safe enough to start with if the alternative is something similar to what Regex would generate for a basic alternation.

Reply:
And if having a ton of alternations like this is a realistic pattern, it might also make sense to let `SearchValues`'s AhoCorasick implementation help.